BetaCameras/Xtion2/Xtion2.cs

```diff
@@ -0,0 +1,182 @@
+        OpenNICAPI.Device _device = new OpenNICAPI.Device();
+        OpenNICAPI.Stream _colorStream = new OpenNICAPI.Stream();
+        OpenNICAPI.Stream _depthStream = new OpenNICAPI.Stream();
+        OpenNICAPI.Stream _irStream = new OpenNICAPI.Stream();
+        OpenNICAPI.Frame _colorFrame = new OpenNICAPI.Frame();
+        OpenNICAPI.Frame _depthFrame = new OpenNICAPI.Frame();
+        OpenNICAPI.Frame _irFrame = new OpenNICAPI.Frame();
+
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (_disposed)
+                return;
+
+            OpenNICAPI.Shutdown();
+            _disposed = true;
+        }
+
+        #region Constructor
+        public Xtion2()
+        {
+            OpenNICAPI.Init();
+        }
+
+        ~Xtion2()
+        {
+            Dispose(false);
+        }
+        #endregion
+
+        #region MetriCam2 Camera Interface Methods
+
+        protected override void LoadAllAvailableChannels()
+        {
+            ChannelRegistry cr = ChannelRegistry.Instance;
+            Channels.Clear();
+
+            Channels.Add(cr.RegisterChannel(ChannelNames.ZImage));
+            Channels.Add(cr.RegisterChannel(ChannelNames.Color));
+			Channels.Add(cr.RegisterChannel(ChannelNames.Intensity));
+        }
+
+
+        protected override void ConnectImpl()
+        {
+            _device = OpenNICAPI.OpenDevice(this.SerialNumber);
+            _colorStream = OpenNICAPI.CreateStream(_device, OpenNICAPI.OniSensorType.COLOR);
+            _depthStream = OpenNICAPI.CreateStream(_device, OpenNICAPI.OniSensorType.DEPTH);
+            _irStream = OpenNICAPI.CreateStream(_device, OpenNICAPI.OniSensorType.IR);
+
+            ActivateChannel(ChannelNames.Color);
+        }
+
+        protected override void DisconnectImpl()
+        {
+            OpenNICAPI.DestroyStream(_colorStream);
+            OpenNICAPI.DestroyStream(_depthStream);
+            OpenNICAPI.DestroyStream(_irStream);
+
+            OpenNICAPI.CloseDevice(_device);
+        }
+
+        protected override void UpdateImpl()
+        {
+            if(IsChannelActive(ChannelNames.Color))
+            {
+                OpenNICAPI.ReleaseFrame(_colorFrame);
+                _colorFrame = OpenNICAPI.ReadFrame(_colorStream);
+            }
+
+            if (IsChannelActive(ChannelNames.ZImage))
+            {
+                OpenNICAPI.ReleaseFrame(_depthFrame);
+                _depthFrame = OpenNICAPI.ReadFrame(_depthStream);
+            }
+
+            if (IsChannelActive(ChannelNames.Intensity))
+            {
+                OpenNICAPI.ReleaseFrame(_irFrame);
+                _irFrame = OpenNICAPI.ReadFrame(_irStream);
+            }
+        }
+
+        protected override CameraImage CalcChannelImpl(string channelName)
+        {
+            if (channelName == ChannelNames.Color)
```

Comment:
What happens if the channel is not active, or has never been active, or if Update has not been called before?
(I guess MC2 is in general not very good at handling these cases, so we should add test cases)

Reply:
fixed